### PR TITLE
[MI-3284] Created API to disconnect a user account

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**

### CODE REVIEW

File: server/api.go [hunk 1]
```diff
@@ -303,6 +304,34 @@ func (a *API) connect(w http.ResponseWriter, r *http.Request) {
...
 
+func (a *API) disconnect(w http.ResponseWriter, r *http.Request) {
+	if r.Method == http.MethodOptions {
+		return
+	}
+
+	userID := r.Header.Get("Mattermost-User-ID")
+	teamsUserID, err := a.p.store.MattermostToTeamsUserID(userID)
+	if err != nil {
+		a.p.API.LogError("The account is not connected.", "UserID", userID, "Error", err.Error())
```
Comment (ayusht2810) [inline server/api.go:315]: i think we should change the errors here to something like `unable to get Teams user ID from Mattermost user ID`

File: server/api.go [hunk 2]
```diff
@@ -303,6 +304,34 @@ func (a *API) connect(w http.ResponseWriter, r *http.Request) {
...
+		http.Error(w, "The account is not connected.", http.StatusBadRequest)
```
Comment (ayusht2810) [inline server/api.go:316]: i think this should be an internal server error

File: server/api.go [hunk 3]
```diff
@@ -303,6 +304,34 @@ func (a *API) connect(w http.ResponseWriter, r *http.Request) {
...
+		return
+	}
+
+	if _, err = a.p.store.GetTokenForMattermostUser(userID); err != nil {
+		a.p.API.LogError("The account is not connected.", "UserID", userID, "Error", err.Error())
+		http.Error(w, "The account is not connected.", http.StatusBadRequest)
+		return
+	}
```
Comment (ayusht2810) [inline server/api.go:324]: i think this check should be present before the above check

File: server/api_test.go [hunk 4]
```diff
@@ -742,3 +742,84 @@ func TestNeedsConnect(t *testing.T) {
...
+		EnabledTeams          string
+		ExpectedResult        string
+		ExpectedStatusCode    int
+	}{
+		{
+			Name:        "Disconnect: user successfully disconnected",
+			SetupPlugin: func(api *plugintest.API) {},
+			SetupStore: func(store *storemocks.Store) {
+				store.On("MattermostToTeamsUserID", testutils.GetUserID()).Return(testutils.GetID(), nil).Times(1)
+				store.On("GetTokenForMattermostUser", testutils.GetUserID()).Return(nil, nil).Times(1)
```
Comment (ayusht2810) [inline server/api_test.go:761]: if the token returned is nil here, isn't the user disconnected already?

File: server/api_test.go [hunk 5]
```diff
@@ -742,3 +742,84 @@ func TestNeedsConnect(t *testing.T) {
...
+				store.On("SetUserInfo", testutils.GetUserID(), testutils.GetID(), (*oauth2.Token)(nil)).Return(nil).Times(1)
+			},
+			ExpectedResult:     "Your account has been disconnected.",
+			ExpectedStatusCode: http.StatusOK,
+		},
+		{
+			Name: "Disconnect: could not find the Teams ID",
```
Comment (ayusht2810) [inline server/api_test.go:768]:
```suggestion
			Name: "Disconnect: could not find the Teams userID",
```

File: server/api_test.go [hunk 6]
```diff
@@ -742,3 +742,84 @@ func TestNeedsConnect(t *testing.T) {
...
+			SetupPlugin: func(api *plugintest.API) {
+				api.On("LogError", "The account is not connected.", "UserID", testutils.GetUserID(), "Error", "could not find the Teams ID").Once()
```
Comment (ayusht2810) [inline server/api_test.go:770]: same

File: server/api_test.go [hunk 7]
```diff
@@ -742,3 +742,84 @@ func TestNeedsConnect(t *testing.T) {
...
+			},
+			SetupStore: func(store *storemocks.Store) {
+				store.On("MattermostToTeamsUserID", testutils.GetUserID()).Return(testutils.GetID(), errors.New("could not find the Teams ID")).Times(1)
```
Comment (ayusht2810) [inline server/api_test.go:773]:
```suggestion
				store.On("MattermostToTeamsUserID", testutils.GetUserID()).Return("", errors.New("could not find the Teams ID")).Times(1)
```

File: server/api_test.go [hunk 8]
```diff
@@ -742,3 +742,84 @@ func TestNeedsConnect(t *testing.T) {
...
+			},
+			ExpectedResult:     "Error occurred while disconnecting the user.\n",
+			ExpectedStatusCode: http.StatusInternalServerError,
+		},
+	} {
+		t.Run(test.Name, func(t *testing.T) {
+			assert := assert.New(t)
+			plugin := newTestPlugin(t)
+			plugin.configuration.EnforceConnectedUsers = test.EnforceConnectedUsers
+			plugin.configuration.EnabledTeams = test.EnabledTeams
```
Comment (ayusht2810) [inline server/api_test.go:808]: use of this?

File: server/api.go [hunk 9]
```diff
@@ -303,6 +304,34 @@ func (a *API) connect(w http.ResponseWriter, r *http.Request) {
...
+
+	if err = a.p.store.SetUserInfo(userID, teamsUserID, nil); err != nil {
+		a.p.API.LogError("Error occurred while disconnecting the user.", "UserID", userID, "Error", err.Error())
+		http.Error(w, "Error occurred while disconnecting the user.", http.StatusInternalServerError)
+		return
+	}
+
+	_, _ = w.Write([]byte("Your account has been disconnected."))
```
[the remark this answers is not in the document]
Comment (raghavaggarwal2308) [reply server/api.go:332]: I don't think there is a need for this

File: server/api.go [hunk 10]
```diff
@@ -30,18 +34,19 @@ func NewAPI(p *Plugin, store store.Store) *API {
...
-	router.HandleFunc("/lifecycle", api.processLifecycle).Methods("POST")
-	router.HandleFunc("/autocomplete/teams", api.autocompleteTeams).Methods("GET")
-	router.HandleFunc("/autocomplete/channels", api.autocompleteChannels).Methods("GET")
-	router.HandleFunc("/needsConnect", api.needsConnect).Methods("GET", "OPTIONS")
-	router.HandleFunc("/connect", api.connect).Methods("GET", "OPTIONS")
-	router.HandleFunc("/oauth-redirect", api.oauthRedirectHandler).Methods("GET", "OPTIONS")
+	router.HandleFunc("/avatar/{userId:.*}", api.getAvatar).Methods(http.MethodGet)
+	router.HandleFunc("/changes", api.processActivity).Methods(http.MethodPost)
+	router.HandleFunc("/lifecycle", api.processLifecycle).Methods(http.MethodPost)
+	router.HandleFunc("/autocomplete/teams", api.autocompleteTeams).Methods(http.MethodGet)
```
Comment (Kshitij-Katiyar) [inline server/api.go:40]: sub-router for autocomplete

File: server/api_test.go [hunk 11]
```diff
@@ -745,40 +745,39 @@ func TestNeedsConnect(t *testing.T) {
...
 			},
 			ExpectedResult:     "Your account has been disconnected.",
 			ExpectedStatusCode: http.StatusOK,
 		},
 		{
-			Name: "Disconnect: could not find the Teams ID",
+			Name: "Disconnect: could not find the Teams user ID",
 			SetupPlugin: func(api *plugintest.API) {
-				api.On("LogError", "The account is not connected.", "UserID", testutils.GetUserID(), "Error", "could not find the Teams ID").Once()
+				api.On("LogError", "Unable to get Teams user ID from Mattermost user ID.", "UserID", testutils.GetUserID(), "Error", "could not find the Teams ID").Once()
```
Comment (ayusht2810) [inline server/api_test.go:768]:
```suggestion
				api.On("LogError", "Unable to get Teams user ID from Mattermost user ID.", "UserID", testutils.GetUserID(), "Error", "could not find the Teams user ID").Once()
```

File: server/api.go [hunk 12]
```diff
@@ -309,17 +314,17 @@ func (a *API) disconnect(w http.ResponseWriter, r *http.Request) {
...
 	}
 
-	userID := r.Header.Get("Mattermost-User-ID")
-	teamsUserID, err := a.p.store.MattermostToTeamsUserID(userID)
-	if err != nil {
-		a.p.API.LogError("The account is not connected.", "UserID", userID, "Error", err.Error())
+	userID := r.Header.Get(HeaderMattermostUserID)
+	if _, err := a.p.store.GetTokenForMattermostUser(userID); err != nil {
+		a.p.API.LogError("Unable to get the oauth token for the user.", "UserID", userID, "Error", err.Error())
 		http.Error(w, "The account is not connected.", http.StatusBadRequest)
```
Comment (ayusht2810) [inline server/api.go]: i think we can create a middleware for this logic as well, as it can be used in other APIs as well.

File: server/api.go [hunk 13]
```diff
@@ -30,19 +34,19 @@ func NewAPI(p *Plugin, store store.Store) *API {
...
-	router.HandleFunc("/disconnect", api.disconnect).Methods("GET", "OPTIONS")
-	router.HandleFunc("/oauth-redirect", api.oauthRedirectHandler).Methods("GET", "OPTIONS")
+	router.HandleFunc("/avatar/{userId:.*}", api.getAvatar).Methods(http.MethodGet)
+	router.HandleFunc("/changes", api.processActivity).Methods(http.MethodPost)
+	router.HandleFunc("/lifecycle", api.processLifecycle).Methods(http.MethodPost)
+	router.HandleFunc("/autocomplete/teams", api.autocompleteTeams).Methods(http.MethodGet)
+	router.HandleFunc("/autocomplete/channels", api.autocompleteChannels).Methods(http.MethodGet)
+	router.HandleFunc("/needsConnect", api.needsConnect).Methods(http.MethodGet, http.MethodOptions)
+	router.HandleFunc("/connect", api.connect).Methods(http.MethodGet, http.MethodOptions)
+	router.HandleFunc("/disconnect", api.handleAuthRequired(api.disconnect)).Methods(http.MethodGet, http.MethodOptions)
```
Comment (ayusht2810) [inline server/api.go:44]: don't we need to apply this middleware in other APIs as well?

File: server/api.go [hunk 14]
```diff
@@ -303,6 +313,32 @@ func (a *API) connect(w http.ResponseWriter, r *http.Request) {
 	_, _ = w.Write(data)
 }
 
+func (a *API) disconnect(w http.ResponseWriter, r *http.Request) {
+	if r.Method == http.MethodOptions {
```
Comment (Kshitij-Katiyar) [inline server/api.go:317]: Why this is needed

**Review by Kshitij-Katiyar — APPROVED** [review]

File: server/api_test.go [hunk 15]
```diff
@@ -742,3 +742,90 @@ func TestNeedsConnect(t *testing.T) {
...
+		{
+			Name: "Disconnect: could not find the Teams user ID",
+			SetupPlugin: func(api *plugintest.API) {
+				api.On("LogError", "Unable to get Teams user ID from Mattermost user ID.", "UserID", testutils.GetUserID(), "Error", "could not find the Teams user ID").Once()
+			},
+			SetupStore: func(store *storemocks.Store) {
+				store.On("GetTokenForMattermostUser", testutils.GetUserID()).Return(&oauth2.Token{}, nil).Times(1)
+				store.On("MattermostToTeamsUserID", testutils.GetUserID()).Return("", errors.New("could not find the Teams user ID")).Times(1)
+			},
+			ExpectedResult:     "The account is not connected.\n",
```
Comment (manojmalik20) [inline server/api_test.go:774]: This is not matching with the API.

**Review by manojmalik20 — COMMENTED** [review]
Please get the PR approved by @ayusht2810 first.

**Review by manojmalik20 — APPROVED** [review]

**Pull request merged**